finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala

```diff
@@ -389,6 +389,17 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
    */
   def send(msg: SendMsg): Future[Future[Unit]] = {
 
+    // if we receive a response with an error status,
+    // it is important to cancel the stream of the rcvMessage
+    // since there might be frames there that will never make
+    // it to the remote and will cause a memory leak...
+    def onErrorResponseWritten(): Unit = msg match {
```

Comment:
I'm a bit confused about this.  `msg` is the message to be sent to the remote, not a message received from the remote.  Why do we match on the send message?

Reply:
So lets say we have a `client -> linkerd -> server` situation. The server is unresponsive, cannot be reached. Whats happening is that in this case the `ErrorReseter` will actually catch that `RichConnectionFailedExceptionWithPath` and translate it to `Future.value(LinkerdHeaders.Err.respond(e.exceptionMessage, Status.BadGateway)): Response` to be sent to the remote. In this case the remote is the client. Now, this is all fine because the client will get a proper description of what happened along with the appropriate status code. However this transport may still have frames in its recvQ that never made it to the server. We need to release these. This is why I match on the message that is being sent back to the client and if its indicating an error, I clear out the recvQ. At least this is my understanding. I might be completely wrong here. Let me know :)

Reply:
Gotcha.  I think we can solve this in a more robust way elsewhere.  It seems that the problem is that when a connection cannot be established, the request that triggered that connection is dropped without draining the request's stream.  Take a look at RoutingFactory.scala.  It seems like perhaps if the call to `clientFactory` fails, the request should be properly disposed of, instead of just dropped.  

This is somewhat complicated by the fact that the request type is generic but the request stream discard behavior is protocol specific.  We deal with a similar problem elsewhere by using a ResponseDiscarder.  We can probably do something similar here and have a RequestDiscarder which is passed to this module as a stack param.

Reply:
Hmmm, yes I will take a closer look into that as I have the feeling that the  `Netty4StreamTransport` is probably not the best place to intercept that.